File: 2022/OEPE/0121.LeetCode.20.ValidParentheses/source.py
_BRACKETS_PAIRS_ = {"(": ")", "{": "}", "[": "]"}


class Solution:
    def isValid(self, s: str) -> bool:
        brackets = []
        for ch in s:
            close_bracket = _BRACKETS_PAIRS_.get(ch, None)            
            if close_bracket is not None:  # open bracket
                brackets.append(close_bracket)            
            elif len(brackets) == 0 or ch != brackets[-1]:                
                return False
            else:
                brackets.pop()
        return len(brackets) == 0


if __name__ == "__main__":
    print(Solution().isValid("()"))
    print(Solution().isValid("{}"))
    print(Solution().isValid("[]"))

File: Archived/2022/OEPE/0121.LeetCode.20.ValidParentheses/source.cpp
#include <string>
#include <iostream>
#include <stack>

using namespace std;

const char* OPEN_BRACKETS = "({[";
const char* CLOSE_BRACKETS = ")}]";

class Solution {
public:
    bool isValid(string s) {
        stack<int> brackets;
        for (int i = 0; i < s.length(); i++) {
            for (int j = 0; j < 3; j++) {
                if (s[i] == OPEN_BRACKETS[j]) {
                    brackets.push(j);                    
                } else if (s[i] == CLOSE_BRACKETS[j]) {
                    if (brackets.size() == 0) {
                        return false; // close bracket without open
                    } else {
                        if (brackets.top() != j) {
                            return false; // open bracket not match with close bracket
                        }
                        brackets.pop();
                    }                    
                }
            }
        }
        if (brackets.size() != 0) {
            return false; // open bracket remain
        }
        return true;
    }
};


int main(){
    cout << Solution().isValid("()") << endl;
    cout << Solution().isValid("{}") << endl;
    cout << Solution().isValid("[]") << endl;
    cout << Solution().isValid(")") << endl;
    cout << Solution().isValid("}") << endl;
    cout << Solution().isValid("]") << endl;
    cout << Solution().isValid("(())") << endl;
    cout << Solution().isValid("({[]})") << endl;
    cout << Solution().isValid("({[]}") << endl;
    cout << Solution().isValid("({[})") << endl;
    cout << Solution().isValid("([]})") << endl;
    return 0;
}